File: 2015/day05.c
#include <stdio.h>
#include <string.h>

int vowel3(const char *s) {
    int v = 0;
    while (*s) {
        if (*s == 'a') v++;
        if (*s == 'e') v++;
        if (*s == 'i') v++;
        if (*s == 'o') v++;
        if (*s == 'u') v++;
        s++;
        if (v == 3) break;
    }
    return (v >= 3);
}

int dblltr(const char *s) {
    const char *t = s + 1;
    while (*t) {
        if (*s == *t) return 1;
        s++;
        t++;
    }
    return 0;
}

int abcdpqxy(const char *s) {
    const char *t = s + 1;
    while (*t) {
        if ((*s == 'a') && (*t == 'b')) return 1;
        if ((*s == 'c') && (*t == 'd')) return 1;
        if ((*s == 'p') && (*t == 'q')) return 1;
        if ((*s == 'x') && (*t == 'y')) return 1;
        s++;
        t++;
    }
    return 0;
}

int part1(char (*data)[20], unsigned n) {
    char buf[100];
    int nice = 0;
    for (unsigned k = 0; k < n; k++) {
        strcpy(buf, data[k]);
        if (!vowel3(buf)) continue;
        if (!dblltr(buf)) continue;
        if (abcdpqxy(buf)) continue;
        nice++;
    }
    return nice;
}

int pair(const char *s) {
    char p[3] = "";
    const char *t = s + 1;
    while (*t) {
        p[0] = *s;
        p[1] = *t;
        if (strstr(t + 1, p)) return 1;
        s++;
        t++;
    }
    return 0;
}

int repeat(const char *s) {
    const char *t = s + 2;
    while (*t) {
        if (*s == *t) return 1;
        s++;
        t++;
    }
    return 0;
}

int part2(char (*data)[20], unsigned n) {
    char buf[100];
    int nice = 0;
    for (unsigned k = 0; k < n; k++) {
        strcpy(buf, data[k]);
        if (!pair(buf)) continue;
        if (!repeat(buf)) continue;
        nice++;
    }
    return nice;
}

int main(void) {
    char data[1200][20];
    unsigned ndata = 0;
    while (fgets(data[ndata], 20, stdin)) {
        data[ndata][strlen(data[ndata]) - 1] = 0;
        ndata++;
    }
    printf("DAY05-PART1: There are %d nice strings.\n", part1(data, ndata));
    printf("DAY05-PART2: There are %d nice strings.\n", part2(data, ndata));
    return 0;
}

File: 2015/day05.fs
\ gforth day05.fs -e part# <.input-day05
\ Make sure to add a new empty line to the end of input

require aoc2015.fs

\ == part1 =====================================================
create &inputword 100 allot
: getstring ( -- len )                  \ and &inputword changed
   0 begin key dup newline? invert while
   over &inputword + c! 1+ repeat drop ;

: vowelcount ( len -- n ) 0 swap 0 do &inputword i + c@
   dup [char] a = rot + swap
   dup [char] e = rot + swap
   dup [char] i = rot + swap
   dup [char] o = rot + swap
       [char] u = + loop negate ;
: doubleletter? ( len -- f ) 1 do &inputword i + dup >r
   c@ r> 1 - c@ = if TRUE unloop exit then loop FALSE ;
: ab,cd,pq,xy? ( len -- f ) 1 do
   &inputword i + dup >r c@ r> 1 - c@
   2dup [char] a = swap [char] b = and if 2drop TRUE unloop exit then
   2dup [char] c = swap [char] d = and if 2drop TRUE unloop exit then
   2dup [char] p = swap [char] q = and if 2drop TRUE unloop exit then
        [char] x = swap [char] y = and if       TRUE unloop exit then
   loop FALSE ;

: nice? ( len -- f )
   dup vowelcount 2 <= if drop FALSE then
   dup doubleletter? invert if drop FALSE then
       ab,cd,pq,xy? invert ;

: part1
   0 begin getstring dup 0 <> while
      nice? - repeat drop
   cr ." Q: How many strings are nice?"
   cr ." A: There are " . ." nice strings." cr cr ;

\ == part2 =====================================================
: ppair? ( i -- f ) &inputword + dup >r c@ r> 1 - c@ = ;
: matchpair? ( len i -- f )
   dup &inputword + c@ rot rot ( &[i] len i )
   2 + do
      dup &inputword j + c@ = if
         &inputword j + 1 - c@ = if
            TRUE unloop exit then
         then
      loop drop FALSE ;

: doublepair? ( len -- f )
   dup 2 - 1 do
      i ppair? if
         dup i matchpair? if
            TRUE unloop exit
            then
         then
      loop
   drop FALSE ;
: tuckedletter? ( len -- f) 2 do &inputword i + dup >r
   c@ r> 2 - c@ = if TRUE unloop exit then loop FALSE ;

: nice2? ( len -- f )
   dup doublepair? invert if drop FALSE then
       tuckedletter? ;

: part2
   0 begin getstring dup 0 <> while
      nice2? - repeat drop
   cr ." Q: How many strings are nice under these new rules?"
   cr ." A: There are " . ." new nice strings." cr cr ;
